File: apps/radio-station/server/services/pick.ts
import type {Track, Weights} from '../types.js';
import {pickUniform, pickWeighted, type Rng} from './rng.js';
import {bucketMix, dueScore, type SeparationWindows} from './clock.js';
import type {BucketName, TrackBuckets} from './buckets.js';

const GOLD_2000S_SHARE = 0.6;
const RELAX_STEPS = [0, 0.2, 0.4, 0.6, 0.8, 1];

export function chooseBucket(rng: Rng, weights: Weights): BucketName {
    if (rng() < weights.goldLeak / 100) {
        return rng() < GOLD_2000S_SHARE ? 'gold2000s' : 'gold1990s';
    }
    const mix = bucketMix(weights.hitGravity);
    return pickWeighted(rng, ['power', 'current', 'recurrent'], [mix.power, mix.current, mix.recurrent]);
}

export function passesSeparation(
    track: Track,
    now: number,
    lastByTrack: Map<string, number>,
    lastByArtist: Map<string, number>,
    windows: SeparationWindows,
    relax: number,
    orbitMinMs = 0,
    numberOneId?: string,
): boolean {
    const scale = 1 - relax;
    const artistLast = lastByArtist.get(track.artist);
    if (artistLast != null && now - artistLast < windows.artistMs * scale) return false;
    const titleLast = lastByTrack.get(track.id);
    if (titleLast != null && now - titleLast < windows.titleMs[track.rotation] * scale) return false;
    if (numberOneId && track.id === numberOneId && titleLast != null && now - titleLast < orbitMinMs) return false;
    return true;
}

function candidatesAtRelax(
    pool: Track[],
    now: number,
    lastByTrack: Map<string, number>,
    lastByArtist: Map<string, number>,
    windows: SeparationWindows,
    relax: number,
    orbitMinMs: number,
    numberOneId: string | undefined,
): Track[] {
    return pool.filter((track) =>
        passesSeparation(track, now, lastByTrack, lastByArtist, windows, relax, orbitMinMs, numberOneId),
    );
}

function withoutLastArtist(pool: Track[], lastByArtist: Map<string, number>): Track[] {
    let lastArtist: string | undefined;
    let lastAt = -1;
    for (const [artist, at] of lastByArtist) {
        if (at > lastAt) {
            lastAt = at;
            lastArtist = artist;
        }
    }
    if (!lastArtist || lastAt < 0) return pool;
    const skipped = pool.filter((track) => track.artist !== lastArtist);
    return skipped.length ? skipped : pool;
}

function eligible(
    pool: Track[],
    now: number,
    lastByTrack: Map<string, number>,
    lastByArtist: Map<string, number>,
    windows: SeparationWindows,
    orbitMinMs: number,
    numberOneId: string | undefined,
): Track[] {
    const preferred = withoutLastArtist(pool, lastByArtist);
    for (const relax of RELAX_STEPS) {
        const found = candidatesAtRelax(preferred, now, lastByTrack, lastByArtist, windows, relax, orbitMinMs, numberOneId);
        if (found.length) return found;
    }
    const safe = preferred.filter((track) =>
        passesSeparation(track, now, lastByTrack, lastByArtist, windows, 1, orbitMinMs, numberOneId),
    );
    if (safe.length) return safe;
    const others = pool.filter((track) => track.id !== numberOneId);
    return others.length ? others : pool;
}

function pickByTemperature(rng: Rng, pool: Track[], now: number, lastByTrack: Map<string, number>, temperature: number): Track {
    if (temperature <= 0) {
        let best = pool[0] as Track;
        let bestScore = dueScore(best, now, lastByTrack.get(best.id));
        for (let i = 1; i < pool.length; i++) {
            const track = pool[i] as Track;
            const score = dueScore(track, now, lastByTrack.get(track.id));
            if (score > bestScore) {
                best = track;
                bestScore = score;
            }
        }
        return best;
    }
    const exp = 1 - 0.99 * (temperature / 100);
    const weights = pool.map((track) => Math.pow(Math.max(dueScore(track, now, lastByTrack.get(track.id)), 1), exp));
    return pickWeighted(rng, pool, weights);
}

export function pickFromBucket(
    rng: Rng,
    buckets: TrackBuckets,
    name: BucketName,
    now: number,
    lastByTrack: Map<string, number>,
    lastByArtist: Map<string, number>,
    windows: SeparationWindows,
    temperature: number,
    orbitMinMs = 0,
    numberOneId?: string,
): Track {
    const pool = buckets[name].length ? buckets[name] : fallbackPool(buckets);
    const fit = eligible(pool, now, lastByTrack, lastByArtist, windows, orbitMinMs, numberOneId);
    if (fit.length === 1) return fit[0] as Track;
    if (temperature >= 100) return pickUniform(rng, fit);
    return pickByTemperature(rng, fit, now, lastByTrack, temperature);
}

function fallbackPool(buckets: TrackBuckets): Track[] {
    return [
        ...buckets.power,
        ...buckets.current,
        ...buckets.recurrent,
        ...buckets.gold2000s,
        ...buckets.gold1990s,
    ];
}

File: apps/radio-station/api/src/main/kotlin/radioapi/domain/Pick.kt
package radioapi.domain

private const val GOLD_2000S_SHARE = 0.6
private val RELAX_STEPS = listOf(0.0, 0.2, 0.4, 0.6, 0.8, 1.0)

fun chooseBucket(rng: Rng, weights: Weights): String {
    if (rng.next() < weights.goldLeak / 100.0) {
        return if (rng.next() < GOLD_2000S_SHARE) "gold2000s" else "gold1990s"
    }
    val mix = bucketMix(weights.hitGravity)
    return pickWeighted(rng, listOf("power", "current", "recurrent"), listOf(mix.power, mix.current, mix.recurrent))
}

fun passesSeparation(
    track: Track,
    now: Long,
    lastByTrack: Map<String, Long>,
    lastByArtist: Map<String, Long>,
    windows: SeparationWindows,
    relax: Double,
    orbitMinMs: Double = 0.0,
    numberOneId: String? = null,
): Boolean {
    val scale = 1 - relax
    val artistLast = lastByArtist[track.artist]
    if (artistLast != null && now - artistLast < windows.artistMs * scale) return false
    val titleLast = lastByTrack[track.id]
    val titleWindow = windows.titleMs[track.rotation] ?: 0.0
    if (titleLast != null && now - titleLast < titleWindow * scale) return false
    return !orbitBlocks(track, now, titleLast, orbitMinMs, numberOneId)
}

fun pickFromBucket(
    rng: Rng,
    buckets: TrackBuckets,
    name: String,
    now: Long,
    lastByTrack: Map<String, Long>,
    lastByArtist: Map<String, Long>,
    windows: SeparationWindows,
    temperature: Int,
    orbitMinMs: Double = 0.0,
    numberOneId: String? = null,
): Track {
    val pool = buckets.named(name).ifEmpty { buckets.all() }
    val fit = eligible(pool, now, lastByTrack, lastByArtist, windows, orbitMinMs, numberOneId)
    if (fit.size == 1) return fit.first()
    if (temperature >= 100) return pickUniform(rng, fit)
    return pickByTemperature(rng, fit, now, lastByTrack, temperature)
}

private fun orbitBlocks(
    track: Track,
    now: Long,
    titleLast: Long?,
    orbitMinMs: Double,
    numberOneId: String?,
): Boolean {
    if (numberOneId == null || track.id != numberOneId || titleLast == null) return false
    return now - titleLast < orbitMinMs
}

private fun eligible(
    pool: List<Track>,
    now: Long,
    lastByTrack: Map<String, Long>,
    lastByArtist: Map<String, Long>,
    windows: SeparationWindows,
    orbitMinMs: Double,
    numberOneId: String?,
): List<Track> {
    val preferred = withoutLastArtist(pool, lastByArtist)
    val relaxed = firstRelaxed(preferred, now, lastByTrack, lastByArtist, windows, orbitMinMs, numberOneId)
    if (relaxed.isNotEmpty()) return relaxed
    val others = pool.filter { it.id != numberOneId }
    return others.ifEmpty { pool }
}

private fun firstRelaxed(
    pool: List<Track>,
    now: Long,
    lastByTrack: Map<String, Long>,
    lastByArtist: Map<String, Long>,
    windows: SeparationWindows,
    orbitMinMs: Double,
    numberOneId: String?,
): List<Track> {
    for (relax in RELAX_STEPS) {
        val found = separated(pool, now, lastByTrack, lastByArtist, windows, relax, orbitMinMs, numberOneId)
        if (found.isNotEmpty()) return found
    }
    return separated(pool, now, lastByTrack, lastByArtist, windows, 1.0, orbitMinMs, numberOneId)
}

private fun separated(
    pool: List<Track>,
    now: Long,
    lastByTrack: Map<String, Long>,
    lastByArtist: Map<String, Long>,
    windows: SeparationWindows,
    relax: Double,
    orbitMinMs: Double,
    numberOneId: String?,
): List<Track> = pool.filter { track ->
    passesSeparation(track, now, lastByTrack, lastByArtist, windows, relax, orbitMinMs, numberOneId)
}

private fun withoutLastArtist(pool: List<Track>, lastByArtist: Map<String, Long>): List<Track> {
    val last = lastByArtist.maxByOrNull { it.value } ?: return pool
    val skipped = pool.filter { it.artist != last.key }
    return skipped.ifEmpty { pool }
}

private fun pickByTemperature(
    rng: Rng,
    pool: List<Track>,
    now: Long,
    lastByTrack: Map<String, Long>,
    temperature: Int,
): Track {
    if (temperature <= 0) return mostDue(pool, now, lastByTrack)
    val exp = 1 - 0.99 * (temperature / 100.0)
    val weights = pool.map { track ->
        val score = dueScore(track, now, lastByTrack[track.id]).coerceAtLeast(1.0)
        Math.pow(score, exp)
    }
    return pickWeighted(rng, pool, weights)
}

private fun mostDue(pool: List<Track>, now: Long, lastByTrack: Map<String, Long>): Track {
    var best = pool.first()
    var bestScore = dueScore(best, now, lastByTrack[best.id])
    for (track in pool.drop(1)) {
        val score = dueScore(track, now, lastByTrack[track.id])
        if (score > bestScore) {
            best = track
            bestScore = score
        }
    }
    return best
}
